### PR TITLE
feat: implement getStateRandao

### PR DESCRIPTION
**Pull request by am1r021**
**Motivation**

See #5701.

**Description**

The goal of this PR is to implement the get-state-randao api.

### CODE REVIEW

**Comment by dapplion** [comment]
@scorbajio thank you for your interest contributing! Please note that opening a PR and pushing commits triggers notifications to the entire team. I would recommend:
- close this PR while you are developing it
- if you have questions please ask in our discord at the #lodestar-general channel, or post the question in the issue https://github.com/ChainSafe/lodestar/issues/5701, we'll answer gladly!
- once your PR is functionally complete and passes tests, re-open the PR as ready to review

File: packages/api/src/beacon/routes/beacon/state.ts [hunk 1]
```diff
@@ -143,6 +143,16 @@ export type Api = {
     >
   >;
 
+  getStateRandao(
```
Comment (nflaig) [inline packages/api/src/beacon/routes/beacon/state.ts]: Might not be ideal to put that method in the middle of  `getStateValidator`-prefixed APIs. Maybe group with `getStateRoot` and `getStateFork`?

File: packages/beacon-node/src/api/impl/beacon/state/index.ts [hunk 2]
```diff
@@ -127,6 +131,42 @@ export function getBeaconStateApi({
...
+        return ret;
+      } else if (
+        usedEpoch < stateEpoch &&
+        Math.abs(stateEpoch - epochsPerHistoricalVector) > 0 &&
+        Math.abs(stateEpoch - epochsPerHistoricalVector) >= usedEpoch
+      ) {
+        return ret;
+      }
+
+      ret.data.randao = bytesToHex(getRandaoMix(state, usedEpoch));
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/index.ts]: You don't need to serialize the data here, will be handled by `RandaoContainer` in `getReturnTypes`

File: packages/beacon-node/src/api/impl/beacon/state/index.ts [hunk 3]
```diff
@@ -127,6 +131,42 @@ export function getBeaconStateApi({
...
+     * Retrieves the sync committees for the given state.
+     * @param epoch Fetch sync committees for the given epoch. If not present then the sync committees for the epoch of the state will be obtained.
+     */
+    async getStateRandao(stateId, epoch) {
+      const {state, executionOptimistic} = await resolveStateId(chain, stateId);
+      const stateEpoch = computeEpochAtSlot(state.slot);
+      const usedEpoch = epoch ?? stateEpoch;
+
+      // TODO is there a way to calculate this by at a specific epoch?
+      const epochsPerHistoricalVector = EPOCHS_PER_HISTORICAL_VECTOR;
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/index.ts]: That shouldn't be required, `getRandaoMix` considers the `EPOCHS_PER_HISTORICAL_VECTOR` already

https://github.com/ChainSafe/lodestar/blob/618895c92922658c77284f9a31273ade7f894e94/packages/state-transition/src/util/seed.ts#L158-L160

File: packages/beacon-node/src/api/impl/beacon/state/index.ts [hunk 4]
```diff
@@ -127,6 +131,42 @@ export function getBeaconStateApi({
...
+
+      const ret = {
+        executionOptimistic,
+        // TODO how to compute finalized?
+        // finalized: state.finalized,
+        data: {
+          randao: "", // return empty value to denote out-of-bound lookup
+        },
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/index.ts]: I think we should throw a 400 error if the provided `epoch` is out of range

~~`getRandaoMix` might not return a value if epoch is out of range, type seems wrong but haven't tested it yet~~ does not throw an error, need to explicitly have a check before

https://github.com/ChainSafe/lodestar/blob/618895c92922658c77284f9a31273ade7f894e94/packages/state-transition/src/util/seed.ts#L159

File: packages/beacon-node/src/api/impl/beacon/state/index.ts [hunk 5]
```diff
@@ -127,6 +131,42 @@ export function getBeaconStateApi({
...
+      const {state, executionOptimistic} = await resolveStateId(chain, stateId);
+      const stateEpoch = computeEpochAtSlot(state.slot);
+      const usedEpoch = epoch ?? stateEpoch;
+
+      // TODO is there a way to calculate this by at a specific epoch?
+      const epochsPerHistoricalVector = EPOCHS_PER_HISTORICAL_VECTOR;
+
+      const ret = {
+        executionOptimistic,
+        // TODO how to compute finalized?
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/index.ts]: `finalized` field is currently not supported by Lodestar and can be ignored for now.

There is a existing issue to tackle this for all existing APIs at once
- https://github.com/ChainSafe/lodestar/issues/5693

File: packages/beacon-node/src/api/impl/beacon/state/index.ts [hunk 6]
```diff
@@ -127,6 +131,42 @@ export function getBeaconStateApi({
       };
     },
 
+    /**
+     * Retrieves the sync committees for the given state.
+     * @param epoch Fetch sync committees for the given epoch. If not present then the sync committees for the epoch of the state will be obtained.
+     */
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/index.ts]: jsdoc needs to be added to interface definition in `packages/api/src/beacon/routes/beacon/state.ts` (see other routes)

File: packages/beacon-node/src/api/impl/beacon/state/index.ts [hunk 7]
```diff
@@ -127,6 +131,42 @@ export function getBeaconStateApi({
...
+
+      if (usedEpoch > stateEpoch) {
+        return ret;
+      } else if (
+        usedEpoch < stateEpoch &&
+        Math.abs(stateEpoch - epochsPerHistoricalVector) > 0 &&
+        Math.abs(stateEpoch - epochsPerHistoricalVector) >= usedEpoch
+      ) {
+        return ret;
+      }
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/index.ts]: Not sure about this part, I think we should follow similar implementation as Lighthouse, seems simple and straight forward, see [PR diff](https://github.com/sigp/lighthouse/pull/3571/files#diff-0aec0f89455da3fc1b452417ec63873cd95bd4ee3319e386233535544f47ab30R894-R924)

As per spec [randao.yaml](https://github.com/ethereum/beacon-APIs/blob/197ecffae2b813d94797184ae50387589c601fe7/apis/beacon/states/randao.yaml):
- `400` for invalid state id or epoch, mostly handled by json schema already (needs to be double checked)
- `400` if epoch is out of range, see [spec note](https://github.com/ethereum/beacon-APIs/blob/197ecffae2b813d94797184ae50387589c601fe7/apis/beacon/states/randao.yaml#L54-L56)
- `404` if state not found

You can also check with other routes that load a state + have epoch as optional param for additional examples

File: packages/beacon-node/src/api/impl/beacon/state/index.ts
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/index.ts:1]: There is an existing test suite for state apis

https://github.com/ChainSafe/lodestar/blob/618895c92922658c77284f9a31273ade7f894e94/packages/beacon-node/test/e2e/api/impl/beacon/state/endpoint.test.ts#L11

Would be a plus if could we could add a simple test case for this

**Comment by nflaig** [comment]
@dapplion I have to push back on this

> if you have questions please ask in our discord at the #lodestar-general channel

While I think discord is great for many things like general questions and technical help, I don't think it is a great tool to collaborate and give code related feedback. It is also good to have all the discussion in one place, which I think a PR is perfect place for it, it also has resolvable threads per topic, and allows directly comment on the code which provides much more clarify.

> once your PR is functionally complete and passes tests, re-open the PR as ready to review

I agree that it doesn't make sense to open a PR that has little to no changes that should be implemented but this is definitely not the case here. It good to open a PR asap in my opinion especially as a new contributor to get feedback on the direction of the changes  and not waste time on things that wouldn't even work in the end. That's why draft exists and a PR is a great collaboration tool we should use as much as possible.

> Please note that opening a PR and pushing commits triggers notifications to the entire team

That's not a good reason why a PR should not be opened, there is always the option to unsubscribe, and that's it. We also had PRs open for weeks (or even months) and I think it is a good thing, it especially for new contributors to follow the work of others. Much better learning experience than just seeing the end result.

**Review by nflaig — COMMENTED** [review]
Thanks for looking into this @scorbajio! Some initial feedback / thoughts

File: packages/api/src/beacon/routes/beacon/state.ts (hunk 1 above)
Comment (am1r021) [reply packages/api/src/beacon/routes/beacon/state.ts]: Fixed.

File: packages/beacon-node/src/api/impl/beacon/state/index.ts (hunk 6 above)
Comment (am1r021) [reply packages/beacon-node/src/api/impl/beacon/state/index.ts]: Fixed.

File: packages/beacon-node/src/api/impl/beacon/state/index.ts (hunk 7 above)
Comment (am1r021) [reply packages/beacon-node/src/api/impl/beacon/state/index.ts]: I was using teku's implementation as a reference. After taking a look at lighthouse, it seems like they do range checking in a more simple way in a separate [get-index helper](https://github.com/michaelsproul/lighthouse/blob/e83726acdf1111e5b9309d4e11e19cfafd2c21ff/consensus/types/src/beacon_state.rs#L951), which is what I used to rework the range-checking logic.

**Review by nflaig — APPROVED** [review]
Looks good to me, thanks for the updates @scorbajio

**Review by g11tech — APPROVED** [review]
lgtm

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.12.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.12.0) :tada:

**Pull request merged**